proposals/NNNN-add-customdebugdescription-conformance-to-anykeypath.md

```diff
@@ -0,0 +1,158 @@
+
+## Detailed design
+
+### Implementation of `CustomDebugStringConvertible`
+
+Much like the `_project` functions currently implemented in `KeyPath.swift`, this function would loop through the keypath's buffer, handling each segment as follows:
+
+For offset segments, the implementation is simple: use `_getRecursiveChildCount`, `_getChildOffset`, and `_getChildMetadata` to get the string name of the property. I believe these are the same mechanisms used by `Mirror` today.
+
+For optional chain, force-unwrap, etc. the function appends a hard coded "?" or "!", as appropriate.
+
+For computed segments, call `swift::lookupSymbol()` on the result of `getter()` in the `ComputedAccessorsPtr`. Demangle the result to get the property name. 
+
+### Changes to the Swift Runtime
+
+To implement descriptions for computed segments, it is necessary to make two changes to the runtime: 
+
+1. Expose a Swift calling-convention function to call `swift::lookupSymbol()`.
+2. Implement and expose a function to demangle keypath functions without the ornamentation the existing demangling functions produce. 
+
+### Dealing with missing data
+
+There are two known cases where data might not be available: 
+
+1. type metadata has not been emitted because the target was built using the `swift-disable-reflection-metadata` flag
+2. the linker has stripped the symbol names we're trying to look up
+
+In these cases, we would print the following: 
+
+#### Offset case
+`<offset [x] ([typename])>` where `x` is the memory offset we read from the reflection metadata, and `typename` is the type that will be returned. 
+So 
+```
+print(\Theme.backgroundColor) // outputs "\Theme.<offset 0 (Color)>"
+```
+
+#### `lookupSymbol` failure case
+
+In this case we'll print the address-in-memory as hex, plus the type name: 
+```
+print(\Theme.overlay) // outputs \Theme.<computed 0xABCDEFG (Color)>
+```
+
+As it might be difficult to correlate a memory address with the name of the function, the type name may be useful here to provide extra context. 
+
+## Source compatibility
+
+Programs that extend `AnyKeyPath` to implement `CustomDebugStringConvertible` themselves will no longer compile and the authors of such code will have to delete the conformance. Based on a search of Github, there are currently no publically available Swift projects that do this. 
+
+Calling `print` on a KeyPath will, of course, produce different results than before. 
+
+It is unlikely that any existing Swift program is depending on the existing behavior in a production context. While it is likely that someone, somewhere has written code in unit tests that depends on the output of this function, any issues that result will be easy for the authors of such code to identify and fix, and will likely represent an improvement in the readability of those tests. 
+
+## Effect on ABI stability
+
+This proposal will add a new var & protcol confromance to the Standard Library's ABI. It will be availability guarded appropriately. 
```

Comment:
* "protcol confromance" → "protocol conformance"

```suggestion
This proposal will add a new var & protocol conformance to the Standard Library's ABI. It will be availability guarded appropriately.
```